src/app/classes/diagram/run.ts

```diff
@@ -119,4 +121,86 @@ export class Run {
 
         return check;
     }
+
+    public hasCycles(): boolean {
+        return this.getCycles().length > 0;
+    }
+
+    public removeCycles(): void {
+        this.getCycles().forEach((arc) => {
+            return this.arcs.splice(
+                this.arcs.findIndex((a) => a === arc),
+                1
+            );
+        });
+    }
+
+    /**
+     * check all arcs sequences for cycles
+     * @returns Last arcs in cyclic arc sequences
+     */
+    private getCycles(): Arc[] {
+        const visitedArcs = new Set<Arc>();
+        const cyclicArcs = new Array<Arc>();
+
+        this._arcs.forEach((arc) => {
+            const visitedTransitions = new Set<Element>();
+            if (arc.sourceEl) visitedTransitions.add(arc.sourceEl);
+            this.checkArcCycle(
+                arc,
+                visitedArcs,
+                visitedTransitions,
+                cyclicArcs
+            );
+        });
+        return cyclicArcs;
+    }
+
+    /**
+     * checks an arc sequence for cycles
+     * @param arc starting arc
+     * @param visitedArcs already visited arcs
+     * @param visitedTransitions already visited transition
+     * @param cyclicArcs last arcs when a cycle occurs
+     */
+    private checkArcCycle(
+        arc: Arc,
+        visitedArcs: Set<Arc>,
+        visitedTransitions: Set<Element>,
+        cyclicArcs: Arc[]
+    ): void {
+        if (!visitedArcs.has(arc) && arc.targetEl) {
+            visitedArcs.add(arc);
+
+            //transition already visited in this sequence?
+            if (visitedTransitions.has(arc.targetEl)) {
+                cyclicArcs.push(arc);
+            } else {
+                visitedTransitions.add(arc.targetEl);
+
+                //continue with the sequences
+                arc.targetEl.outogingArcs.forEach((outArc) => {
+                    this.checkArcCycle(
+                        outArc,
+                        visitedArcs,
+                        visitedTransitions,
+                        cyclicArcs
+                    );
+                });
+
+                visitedTransitions.delete(arc.targetEl);
+            }
+        }
```

Comment:
Die Komplexität könnte man ziemlich runterschrauben, in dem du hier mit Guards arbeitest statt mit if-else.

```suggestion
        if (visitedArcs.has(arc) || !arc.targetEl) {
            return;
        }

        visitedArcs.add(arc);

        //transition already visited in this sequence?
        if (visitedTransitions.has(arc.targetEl)) {
            cyclicArcs.push(arc);
            return;
        }

        visitedTransitions.add(arc.targetEl);

        //continue with the sequences
        arc.targetEl.outogingArcs.forEach((outArc) => {
            this.checkArcCycle(
                outArc,
                visitedArcs,
                visitedTransitions,
                cyclicArcs
            );
        });

        visitedTransitions.delete(arc.targetEl);
```